fusion/FSL.md

```diff
@@ -0,0 +1,1220 @@
+ioput.input and ioput.output objects to the same buffer a modification
+in-place idiom can be implemented. This isolates the style used in
+the conversion clause from these external mechanics.
+
+### ioput methods
+
+ioput elements have these methods available. 
+
+```
+.clear()                 empties the associated container
+.erase(index,size)       removes size elements begining with index,
+                         size is inclusive of the index position.
+.insert(index,instr)     insert an instr object before index. 
+                         instr is an FSL type
+.replace(sequence,instr) removes sequence.length objects beginning
+                         at sequence.index, and inserts instr 
+```
+
+## Sequence clause
+
+The sequence clause is used to match the incoming instructions to a
+known transform. 
+
+The arguments to the sequence clause are the input sequence name and
+the Mavis facade name, declared using ioput and isa, respectively.
+
+The ioput and isa objects give the sequence clause access to the 
+instruction containers and known instruction definitions. This
+access is used for validation of the sequence and for matching.
+
+The sequence clause implicitly uses the 1st element of ioput, the 
+input container. The sequence clause does not modify the ioput containers.
+
+```
+ioput iop1
+
+sequence seq1(iop1,myISA) {
+   # sequence of instructions
+}
+```
+
+The FSL parser uses the Fusion API and access to the isa object to
+validate the instruction sequence. An exception is thrown if a problem
+is detected.
+
+The specified sequence of instructions is pattern matched against 
+the contents of the input container.  If a match is detected the sequence
+object, seq1, will update its internal index field and length field. This
+information is passed to the constraints clause. On a match the sequence
+clause returns \_pass\_ to the Fusion API.
+
+The index field is the zero-referenced position of the start of the match 
+within ioput. The length is the number of instructions, inclusive of the
+starting instruction.
+
+An instruction sequence is expressed as a simple list of Mavis-assigned
+unique identifiers, or as a set of assembly language instructions with
+abstracted operands. The choice is based on the constraints.
+
+The UID list is sufficient if there are no constraints on the operands
+in the list of instructions in the sequence.
+
+For example, these two instructions are a frequent sequence in compiled
+C.  The left column is the encoding, the comment is the Mavis unique identifier 
+for each instruction which is independent of the operands.
+
+```
+0x0512     c.slli x10,4    ; 0xf
+0x8111     c.srli x10,4    ; 0x13
+```
+
+If you do not require constraints on the operands and therefore have no
+need to refer to the operands by name, you can simply specify the UIDs 
+as a list, one per line.
+
+```
+ioput iop1
+
+sequence seq1(iop1,rv64g) {
+  0xf
+  0x13
+}
+```
```

Comment:
Using UID style specification will have an impact of readability and may need comments describing the UID.  Is there a reason why we need the UID style specification and not just use the abstract instruction style specification?

Reply:
The UID form lends it's self to automated flows. The UID form comes out of the practical work I'm doing. If I do not need to express operand specific constraints then I can more fully automate the discovery of possible fusion tuples, express these to the model, perform the simulation and extract the results. This is a very low touch methodology.

Having said that UIDs is not the only method. There is support for assembly language expression. "...as a set of assembly language instructions with..."

Comments are possible, there are no comments per UID in this example since the associated instruction examples are shown directly above, line ~610

Once the FSL parser has an integrated RISC-V assembly language parser that supports abstract operands the assembly language form of these examples will be available.